Fix flaky TestTrainingLoop - TestE2ETensorPipe

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51939 Fix flaky TestTrainingLoop - TestE2ETensorPipe**

TestTrainingLoop - TestE2ETensorPipe was flaky since there would still
be inflight background RPCs running as we performed the assertions. This
resulted in these assertions failing since we didn't wait for all RPCs on the
agent to finish.

To resolve this issue, in this PR we join() and shutdown() the RPC agent to
ensure no further RPCs are done. Then we assertion the map sizes to ensure no
leaks occurred.

In addition to this, added messageIdToTimeout map to lookup the appropriate
timeout for a messageId. This ensures we remove the appropriate entry from the
map. The previous solution was passing the expirationTime through the lambda,
but it is not guaranteed the lambda would read the response of the request we
just sent out.

Differential Revision: [D26331585](https://our.internmc.facebook.com/intern/diff/D26331585/)

pritamdamania committed:
Fix flaky TestTrainingLoop - TestE2ETensorPipe

TestTrainingLoop - TestE2ETensorPipe was flaky since there would still
be inflight background RPCs running as we performed the assertions. This
resulted in these assertions failing since we didn't wait for all RPCs on the
agent to finish.

To resolve this issue, in this PR we join() and shutdown() the RPC agent to
ensure no further RPCs are done. Then we assertion the map sizes to ensure no
leaks occurred.

In addition to this, added messageIdToTimeout map to lookup the appropriate
timeout for a messageId. This ensures we remove the appropriate entry from the
map. The previous solution was passing the expirationTime through the lambda,
but it is not guaranteed the lambda would read the response of the request we
just sent out.

Differential Revision: [D26331585](https://our.internmc.facebook.com/intern/diff/D26331585/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -958,7 +965,7 @@ std::shared_ptr<JitFuture> TensorPipeAgent::send(
         pipeRead(
             clientPipe.pipe_,
-            [this, &clientPipe, expirationTime](
+            [this, &clientPipe, messageIdR = messageId](
```
So IIUC, this means that it's not necessary that `pipeRead` here is guaranteed to read the response that was sent in the associated `pipeWrite` above?

pritamdamania87 replied:
Yes, I think this is by design of tensorpipe. I think the pipeRead is just called to ensure we fire a "read" request to process an incoming response, but it doesn't necessarily have to be the response for the same request. It will just read whatever it receives on the wire.

rohan-varma reviewed (approved):
LGTM, although there is a valid lint error

pritamdamania committed:
Update on "Fix flaky TestTrainingLoop - TestE2ETensorPipe"


TestTrainingLoop - TestE2ETensorPipe was flaky since there would still
be inflight background RPCs running as we performed the assertions. This
resulted in these assertions failing since we didn't wait for all RPCs on the
agent to finish.

To resolve this issue, in this PR we join() and shutdown() the RPC agent to
ensure no further RPCs are done. Then we assertion the map sizes to ensure no
leaks occurred.

In addition to this, added messageIdToTimeout map to lookup the appropriate
timeout for a messageId. This ensures we remove the appropriate entry from the
map. The previous solution was passing the expirationTime through the lambda,
but it is not guaranteed the lambda would read the response of the request we
just sent out.

Differential Revision: [D26331585](https://our.internmc.facebook.com/intern/diff/D26331585/)

[ghstack-poisoned]
